src/graphapi/opgraph.cpp

```diff
@@ -0,0 +1,109 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2023 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/graphapi/opgraph.hpp>
+
+#include <unordered_map>
+
+namespace miopen {
+namespace graphapi {
+
+OpNode::~OpNode() = default;
+
+OpGraph OpGraphBuilder::build() &&
+{
+
+    OpGraph graph;
+
+    // key = tensor ptr, value = vec. of dest nodes
+    /// \todo might eventually move this state to the graph class during build()
+    std::unordered_map<Tensor*, EdgeInfo> e_map;
+
+    for(OpNode* n : mNodes)
+    {
+
+        for(Tensor* i : n->getInTensors())
+        {
+            auto [iter, _ignore] =
+                e_map.try_emplace(i, EdgeInfo{}); // add empty EdgeInfo if not present
+
+            iter->second.mDests.emplace_back(n);
+        }
+
+        for(Tensor* o : n->getOutTensors())
+        {
+            auto [iter, _ignore] =
+                e_map.try_emplace(o, EdgeInfo{}); // add empty EdgeInfo if not present
+
+            assert(iter->second.mSrc == nullptr);
+            iter->second.mSrc = n;
+        }
+    }
+
+    graph.initNodes(std::move(mNodes));
+
+    for(const auto& [tens_ptr, edge_info] : e_map)
+    {
+
+        if(edge_info.mSrc != nullptr && !edge_info.mDests.empty())
+        {
+            for(const auto& d : edge_info.mDests)
+            {
+                graph.addEdge(edge_info.mSrc, tens_ptr, d);
+            }
+        }
+        else if(edge_info.mSrc == nullptr)
+        {
+
+            // tens_ptr is a non-virtual input tensor
+            assert(!tens_ptr->isVirtual()); // tensor pointer can't be virtual
```

Comment:
```C++
... // tensor pointer can't be virtual
```
Yep, this looks like a misconfigured graph. But do we \<plan to\> guarantee that this condition is met somewhere else? If not, then I would highly recommend converting this to MIOPEN_THROW.

Reply:
Seems it must not be checked.
Input tensors can be some sort of virtual.
For example batchnorm operation consumes a lot of input tensors, but some of them are optional.

Reply:
I will have to think about it. What does it mean for inputs to be virtual? I thought inputs must be provided or you can't compute a valid result but you're right there might be cases when inputs are optional and a valid output can still be produced.

Reply:
I've taken the check out for now. Both input and output tensors can be virtual.

Reply:
> I will have to think about it. What does it mean for inputs to be virtual? I thought inputs must be provided or you can't compute a valid result but you're right there might be cases when inputs are optional and a valid output can still be produced.

The primitive just ignores some inputs, that's it. Either the input can be some precalculated value and primitive calculates it by itself, or it just work in another mode without those tensors.
There are two options:
1) the user does not provide those tensors to the primitive at all - we won't see any input virtual tensors, but we have to feed nullpointers internally
2) the user explicitly marks those tensors virtual

Reply:
yeah, it's possible that some operations don't need all inputs, in which case the user must mark them `virtual` in my opinion.

Reply:
[resolved]